src/test/crimson/test_messenger_thrash.cc

```diff
@@ -450,9 +450,9 @@ class SyntheticWorkload {
      int i = 0;
      return seastar::do_until(
        [this] { return !dispatcher.get_num_pending_msgs(); },
-       [this, &i]
```

Comment:
Thanks for the fix!

The issue is valid, but `i` needs to be increased below -- it cannot be captured by copy -- it should be captured by reference (as it is).

How about instead wrapping `i` with `seastar::do_with()` to make sure it is alive during `seastar::do_until()`?

Reply:
for "[this, i] () mutable", the lambda function object just takes the value of "i", but not the address of "i".
when the return value of dispatcher.get_num_pending_msgs is not zero,  seastar::do_until will try until dispatcher.get_num_pending_msgs returns zero or reactor needs to preempt.
if reactor needs to preempt, seastar::do_until will create a async task to check pending msgs, the async task takes the value of "i" of the lambda function object too.

Reply:
I think we should not rely on this assumption -- it looks dependent on when the lambda is copied internally inside seastar implementations. And I recall that there were unexpected bugs relying on this assumtion, which is hard to be identified as well.

`seastar::do_with()` explicitly carifies the intension and should be always correct, what do you think?

Reply:
yes, seastar::do_with() is more appropriate. the new commit has been updated.